src/main/java/org/sagebionetworks/bridge/hibernate/HibernatePermissionDao.java

```diff
@@ -0,0 +1,101 @@
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+
+import org.sagebionetworks.bridge.dao.PermissionDao;
+import org.sagebionetworks.bridge.models.permissions.Permission;
+import org.sagebionetworks.bridge.models.permissions.EntityType;
+import org.springframework.stereotype.Component;
+
+import javax.annotation.Resource;
+
+import java.util.List;
+import java.util.Optional;
+
+@Component
+public class HibernatePermissionDao implements PermissionDao {
+    
+    static final String APP_ID = "appId";
+    static final String USER_ID = "userId";
+    static final String ENTITY_TYPE = "entityType";
+    static final String ENTITY_ID = "entityId";
+    
+    static final String GET_BY_USER = "FROM Permission WHERE appId=:appId AND userId=:userId";
+    static final String GET_BY_ENTITY_TYPE = "FROM Permission WHERE appId=:appId "+
+            "AND entityType=:entityType AND entityId=:entityId";
+    
+    private HibernateHelper hibernateHelper;
+    
+    @Resource(name = "mysqlHibernateHelper")
+    final void setHibernateHelper(HibernateHelper hibernateHelper) {
+        this.hibernateHelper = hibernateHelper;
+    }
+    
+    @Override
+    public Optional<Permission> getPermission(String appId, String guid) {
+        checkArgument(isNotBlank(appId));
+        checkArgument(isNotBlank(guid));
+        
+        Permission permission = hibernateHelper.getById(Permission.class, guid);
+        if (permission == null) {
+            return Optional.empty();
+        }
+        
+        return Optional.of(permission);
+    }
+    
+    @Override
+    public List<Permission> getPermissionsForUser(String appId, String userId) {
+        checkArgument(isNotBlank(appId));
+        checkArgument(isNotBlank(userId));
+        
+        QueryBuilder builder = new QueryBuilder();
+        builder.append(GET_BY_USER, APP_ID, appId, USER_ID, userId);
+        
+        return hibernateHelper.queryGet(
+                builder.getQuery(), builder.getParameters(), null, null, Permission.class);
+    }
+    
+    @Override
+    public List<Permission> getPermissionsForEntity(String appId, EntityType entityType, String entityId) {
+        checkArgument(isNotBlank(appId));
+        checkNotNull(entityType);
+        checkArgument(isNotBlank(entityId));
+        
+        QueryBuilder builder = new QueryBuilder();
+        builder.append(GET_BY_ENTITY_TYPE, APP_ID, appId, ENTITY_TYPE, entityType, ENTITY_ID, entityId);
+    
+        return hibernateHelper.queryGet(
+                builder.getQuery(), builder.getParameters(), null, null, Permission.class);
+    }
+    
+    @Override
+    public Permission createPermission(String appId, Permission permission) {
+        checkArgument(isNotBlank(appId));
+        checkNotNull(permission);
+    
+        hibernateHelper.create(permission);
```

Comment:
What happens if the permission already exists with the same guid? Does the database throw an error that you translate correctly into an EntityAlreadyExists exception?

Reply:
I didn't look at the guid duplication before because it's generated in the service. But it returns a ConstraintViolationException that says "Cannot update this permission because it has duplicate primary keys". If the guid is unique but there's already a permission with matching user, accessLevel, and entity, then the message changes to "Cannot update this permission because it has duplicate permission records". That is from the existing unique key constraint message in the MySQL hibernate converter.

Reply:
An EntityAlreadyExists exception would be a bit clearer. Should I just check to see if the permission already exists here or in the service before actually making the create() call?

Reply:
Yes it needs to return an EAEE but you don't have to do it here. Either the persistence exception converter can do it, or you can check and throw in the service. As long as the API behavior is the right exception returned, you are good.